File: cpp/cpp_100-999/743.cc
// C++
// leetcode 743
// https://leetcode-cn.com/problems/network-delay-time/
// 迪杰斯特拉 Dijkstra

class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        vector< vector<int> >   time(n + 1, vector<int>(n+1, INT_MAX));
        vector<int> dist(n + 1, INT_MAX);
        vector<int> pre(n + 1, 0);
        set<int>    undone;
        for(int i = 1; i <= n; ++i) {
            undone.insert(i);
        }
        dist[k] = 0;
        pre[k] = k;
        // 把 times 优化一下，装进map
        
        for(auto tri : times) {
            time[ tri[0] ][ tri[1] ] = tri[2];
        }
        int res = -1;
        int next = k;
        int doneNum = 1;
        while(doneNum < n) {
            // 每轮确定一个点
            undone.erase(next);
            {
                // 先把上一个点能刷的刷了
                auto it = undone.begin();
                while(it != undone.end()) {
                    int one = *it;
                    ++it;
                    int newD = time[next][one];
                    if(newD == INT_MAX)
                        continue;
                    newD += dist[next];
                    if(newD < dist[one]) {
                        dist[one] = newD;
                        pre[one] = next;
                    }
                }
                // 刷完了
            }
            int m_loop = INT_MAX;
            {
                // 找下一个点
                auto it = undone.begin();
                while(it != undone.end()) {
                    int one = *it;
                    ++it;
                    if(dist[one] < m_loop) {
                        m_loop = dist[one];
                        next = one;
                    }
                }
            }
            if(m_loop == INT_MAX)
                return  -1;
            if(m_loop > res)
                res = m_loop;
            ++doneNum;
        }
        return  res;
    }
};

File: Go/Go_100-999/743.go
// Go
// leetcode 743
// https://leetcode.cn/problems/network-delay-time/

func networkDelayTime(times [][]int, n int, k int) int {
	// 邻接矩阵法
	mat := make([][]int, n)
	FAR := 1000_000
	for i := 0; i <n; i++ {
		mat[i] = make([]int, n)
		for j:=0; j < n; j++ {
			mat[i][j] = FAR
		}
	}

	for i:=0; i<len(times); i++ {
		a, b, c := times[i][0]-1, times[i][1]-1, times[i][2]
		mat[a][b] = c
	}

	vis := make([]bool, n)
	dis := make([]int, n)
	for i:=0; i<n; i++ {
		dis[i] = FAR
	}

	dis[k-1] = 0
	got := 0
	for got < n {
		goodIdx:= -1
		for i:=0; i<n; i++ {
			if vis[i] || dis[i] == FAR {
				continue
			}
			if goodIdx == -1 {
				goodIdx = i
			} else if dis[i] < dis[goodIdx] {
				goodIdx = i
			}
		}
		if goodIdx == -1 { // 无法到达更多结点
			return -1
		}
		vis[goodIdx] = true
		got++
		if got == n {
			return dis[goodIdx]
		}
		for i:=0; i<n; i++ {
			if vis[i] || mat[goodIdx][i] == FAR {
				continue
			}
			dis[i] = min(dis[i], dis[goodIdx] + mat[goodIdx][i])
		}
	}
	return -2 // never
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}